ghga_service_commons/api/mock_router.py

```diff
@@ -0,0 +1,386 @@
+    an error if a given test doesn't hit every mocked endpoint.
+    """
+    return False
+
+
+class RegisteredEndpoint(BaseModel):
+    """Endpoint data with the url turned into regex string to get parameters in path"""
+
+    url_pattern: str
+    endpoint_function: Callable
+    signature_parameters: dict[str, Any]
+
+
+class MockRouter:
+    """
+    A class used to register mock endpoints with decorators similar to FastAPI.
+
+    Tag endpoint functions with EndpointHandler.[method]("/some/url-with/{variables}").
+    The regex compiler function will turn the url specified in the decorator function
+    into a regex string capable of capturing the variables in the url (curly brackets)
+    with named groups. That in turn enables linking the named path variables to the
+    variables in the endpoint function itself.
+
+    The only parameter types allowed in the endpoint functions are primitives
+    that can be stored in the url string: int, float, str, bool, None, and complex.
+    The one exception is "request", which will be passed in automatically if specified.
+    """
+
+    def __init__(
+        self,
+        http_exception_handler: Optional[
+            Callable[[httpx.Request, HttpException], Any]
+        ] = None,
+    ):
+        """Initialize the MockRouter with an optional HttpException handler.
+
+        Args:
+            http_exception_handler:
+                custom exception handler function that takes the request and exception
+                as arguments, in that order.
+        """
+
+        self.http_exception_handler: Optional[
+            Callable[[httpx.Request, HttpException], Any]
+        ] = http_exception_handler
+
+        self._methods: dict[str, list[RegisteredEndpoint]] = {
+            "GET": [],
+            "DELETE": [],
+            "POST": [],
+            "PATCH": [],
+            "PUT": [],
+        }
+
+    @staticmethod
+    def _ensure_all_parameters_are_typed(
+        endpoint_function: Callable, signature_parameters: dict[str, Any]
+    ):
+        """Verify that all the endpoint function parameters are typed.
+
+        This will not apply to the request parameter because we don't perform any
+        type conversion on that.
+
+        Args:
+            endpoint_function: the function associated with the endpoint.
+            signature_parameters:
+                A dict containing type information for the endpoint function's parameters.
+
+        Raises:
+            TypeError: When one or more parameters are missing type-hint information.
+        """
+
+        all_parameters = signature(endpoint_function).parameters
+
+        for parameter in all_parameters:
+            if parameter not in signature_parameters:
+                raise TypeError(
+                    f"Parameter '{parameter}' in '{endpoint_function.__name__}' is "
+                    + "missing a type hint"
+                )
+
+    @staticmethod
+    def _ensure_decorator_matches_endpoint(
```

Comment:
Might be slightly more expressive calling this something like `_ensure_decorator_parameters_match_endpoint`

Reply:
I went with `_ensure_decorator_and_endpoint_parameters_match` in [2ce6daf](https://github.com/ghga-de/ghga-service-commons/pull/18/commits/2ce6daf24f2699069bfb710c1137dc84cbc9d523)